### PR TITLE
feat!: Add validity timestamps for Relay Calls

### PR DESCRIPTION
**Pull request by b00ste**
# What does this PR introduce?

## ⚠️ BREAKING CHANGES

LSP6KeyManager interfaceId is changed, `executeRelayCall(..)` & `executeRelayCallBatch(..)` functions have an additional parameter.

## 🚀 Feature

Add validity timestamps, start/end for each Relay Call, verify if it's valid and add to the signature.

### PR Checklist

- [x] Wrote Tests
- [ ] Wrote Documentation
- [x] Ran `npm run linter` (solhint)
- [x] Ran `npm run format` (prettier)
- [x] Ran `npm run build`
- [x] Ran `npm run test`

### CODE REVIEW

File: contracts/LSP6KeyManager/ILSP6KeyManager.sol [hunk 1]
```diff
@@ -58,12 +58,14 @@ interface ILSP6KeyManager is
      * @dev allows anybody to execute given they have a signed message from an executor
      * @param signature bytes32 ethereum signature
      * @param nonce the address' nonce (in a specific `_channel`), obtained via `getNonce(...)`. Used to prevent replay attack
+     * @param validityTimestamps two timestamps combined, first timestamp limits from when the payload can be executed and the second timestamp delimits the end of the validity of the payload
```
Comment (CJ42) [inline contracts/LSP6KeyManager/ILSP6KeyManager.sol:61]:
```suggestion
     * @param validityTimestamps two `uint128` timestamps concatenated, the first timestamp determines from when the payload can be executed, the second timestamp delimits the end of the validity of the payload
```

File: contracts/LSP6KeyManager/LSP6Errors.sol [hunk 2]
```diff
@@ -158,3 +158,13 @@ error CannotSendValueToSetData();
...
  */
 error CallingKeyManagerNotAllowed();
+
+/**
+ * @dev reverts when relay call start timestamp is bigger than the current timestamp
+ */
+error RelayCallNotValidYet();
+
+/**
+ * @dev reverts when relay call end timestamp is smaller than the current timestamp
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Errors.sol:168]: Suggesting alternative description for the Natspec

```suggestion
 * @dev reverts when the period to execute the relay call has expired
```

File: contracts/LSP6KeyManager/LSP6KeyManagerCore.sol [hunk 3]
```diff
@@ -326,6 +339,17 @@ abstract contract LSP6KeyManagerCore is
         // increase nonce after successful verification
         _nonceStore[signer][nonce >> 128]++;
 
+        uint128 startTimestamp = uint128(validityTimestamps >> 128);
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6KeyManagerCore.sol:342]: very nice, clean and readable 👌 ❤️

File: tests/utils/helpers.ts [hunk 4]
```diff
@@ -155,9 +155,66 @@ export function combineAllowedCalls(
...
+    days?: number;
+    weeks?: number;
+  },
+  timeAfter: {
+    seconds?: number;
+    hours?: number;
+    days?: number;
+    weeks?: number;
+  },
+  startingTimestamp?: number,
```
Comment (CJ42) [inline tests/utils/helpers.ts:187]: Not sure I understand.
Is this intended to be using:
- for 1st parameter: either `timeBefore` or `startingTimestamp`?
- for 2nd parameter: either `timeAfter` or `endingTimeStamp`?

File: tests/utils/helpers.ts [hunk 5]
```diff
@@ -155,9 +155,66 @@ export function combineAllowedCalls(
...
 
+function countSeconds(time: {
+  seconds?: number;
+  hours?: number;
+  days?: number;
+  weeks?: number;
+}): number {
+  let totalTime: number = 0;
+
+  if (time.seconds) totalTime += time.seconds * 60;
```
Comment (CJ42) [inline tests/utils/helpers.ts:166]: there must be an easy way in JS to do that with `Date` object.

File: tests/utils/helpers.ts (hunk 4 above)
Comment (b00ste) [reply tests/utils/helpers.ts:187]: Either first 2 params or the last 2 params. First 2 generate timestamp from now with the wanted modifications and the last 2 params are for custom timestamps just converted to hex, glued and returned

File: tests/utils/helpers.ts (hunk 5 above)
Comment (b00ste) [reply tests/utils/helpers.ts:166]: will check that

Comment (b00ste) [reply tests/utils/helpers.ts:166]: The way it is seems to be the easiest way, lmk if you find something else.

File: contracts/LSP6KeyManager/ILSP6KeyManager.sol [hunk 6]
```diff
@@ -58,12 +58,14 @@ interface ILSP6KeyManager is
      * @dev allows anybody to execute given they have a signed message from an executor
      * @param signature bytes32 ethereum signature
      * @param nonce the address' nonce (in a specific `_channel`), obtained via `getNonce(...)`. Used to prevent replay attack
+     * @param validityTimestamps two `uint128` timestamps concatenated, the first timestamp determines from when the payload can be executed, the second timestamp delimits the end of the validity of the payload
```
Comment (YamenMerhi) [inline contracts/LSP6KeyManager/ILSP6KeyManager.sol:61]: Maybe we can mention in the @dev tag about the null (meaning no timestamp)

File: tests/LSP6KeyManager/Relay/ExecuteRelayCall.test.ts [hunk 7]
```diff
@@ -575,10 +637,644 @@ export const shouldBehaveLikeExecuteRelayCall = (
...
+                    validityTimestamps,
+                    calldata
+                  )
+              ).to.be.revertedWithCustomError(
+                context.keyManager,
+                "RelayCallExpired"
+              );
+            });
+          });
+          describe("start timestamp < now > end timestamp", () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/Relay/ExecuteRelayCall.test.ts:699]: needing extra blank line before the `describe` block.

File: tests/LSP6KeyManager/Relay/ExecuteRelayCall.test.ts [hunk 8]
```diff
@@ -575,10 +637,644 @@ export const shouldBehaveLikeExecuteRelayCall = (
...
+
+            startingTimestamp =
+              Date.parse(`Tue Apr 20 ${year} 04:20:00 GMT+0000`) / 1000;
+            endingTimestamp =
+              Date.parse(`Mon Apr 19 ${year} 04:20:00 GMT+0000`) / 1000;
+
+            count++;
+          });
+
+          describe("start timestamp = now > end timestamp", () => {
```
Comment (CJ42) [inline tests/LSP6KeyManager/Relay/ExecuteRelayCall.test.ts:659]: can we write the test with plain English sentences please?
It' hard to review and understand the context as we need to decipher the math symbols with the `=` and `<`

File: tests/LSP6KeyManager/Relay/ExecuteRelayCall.test.ts [hunk 9]
```diff
@@ -575,10 +637,644 @@ export const shouldBehaveLikeExecuteRelayCall = (
...
+              const signature = await signLSP6ExecuteRelayCall(
+                context.keyManager,
+                nonce.toString(),
+                validityTimestamps,
+                signerPrivateKey,
+                value,
+                calldata
+              );
+
+              await time.setNextBlockTimestamp(now);
```
Comment (CJ42) [inline tests/LSP6KeyManager/Relay/ExecuteRelayCall.test.ts:682]: nice! 🚀

**Review by CJ42 — CHANGES_REQUESTED** [review]

File: contracts/LSP6KeyManager/LSP6Errors.sol [hunk 10]
```diff
@@ -158,3 +158,13 @@ error CannotSendValueToSetData();
  * @dev reverts when calling the KeyManager through execute(..)
  */
 error CallingKeyManagerNotAllowed();
+
+/**
+ * @dev reverts when relay call start timestamp is bigger than the current timestamp
+ */
+error RelayCallNotValidYet();
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Errors.sol:165]:
```suggestion
error RelayCallBeforeStartTime();
```

File: contracts/LSP6KeyManager/LSP6KeyManagerCore.sol [hunk 11]
```diff
@@ -326,6 +339,19 @@ abstract contract LSP6KeyManagerCore is
         // increase nonce after successful verification
         _nonceStore[signer][nonce >> 128]++;
 
+        if (validityTimestamps != 0) {
+            uint128 startTimestamp = uint128(validityTimestamps >> 128);
+            uint128 endTimestamp = uint128(validityTimestamps);
+
+            // solhint-disable not-rely-on-time
+            if (startTimestamp > block.timestamp) {
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6KeyManagerCore.sol:347]:
```suggestion
            if (block.timestamp < startTimestamp) {
```

File: contracts/LSP6KeyManager/LSP6KeyManagerCore.sol [hunk 12]
```diff
@@ -326,6 +339,19 @@ abstract contract LSP6KeyManagerCore is
...
+                revert RelayCallNotValidYet();
+            }
+            if (endTimestamp < block.timestamp) {
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6KeyManagerCore.sol:350]:
```suggestion
            if (block.timestamp > endTimestamp) {
```

**Review by CJ42 — APPROVED** [review]

**Pull request merged**